### PR TITLE
[1880] More Foreign Investor logic and capitalization event

### PR DESCRIPTION
**Pull request by bentziaxl**
show FI shares in corp view

add b3 event  -> corporations getting more capital if 5 shares are sold, ongoing effect after

FI connection check and merger of FI and corp.

Train marker to round description

FI stops operating event

### CODE REVIEW

File: lib/engine/game/g_1880/game.rb [hunk 1]
```diff
@@ -146,7 +150,9 @@ class Game < Game::Base
                     price: 180,
                     rusts_on: '6',
                     num: 5,
-                    events: [{ 'type' => 'float_30' }, { 'type' => 'permit_b' }],
+                    events: [{ 'type' => 'float_30' },
```
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:153]: Did you remove the permit_b event purposely or merge error?

File: lib/engine/game/g_1880/game.rb [hunk 2]
```diff
@@ -323,25 +329,70 @@ def init_minors
...
+        end
+
         def p1
           company_by_id('P1')
         end
 
         def status_array(corporation)
-          return [] unless corporation.floated?
+          return unless corporation.floated?
+          return if corporation.minor?
```
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:346]: Make these one statement `return if corporation.minor? || !corporation.floated?`

File: lib/engine/game/g_1880/game.rb (hunk 1 above)
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb:153]: merge error

File: lib/engine/game/g_1880/game.rb [hunk 3]
```diff
@@ -323,25 +329,70 @@ def init_minors
...
 
+        def non_floated_corporations
+          @corporations.each { |c| yield c if !c.minor? && !c.floated? }
+        end
+
         def corporation_show_individual_reserved_shares?(_corporation)
           false
         end
 
+        def event_all_shares_available!
```
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:359]: Do you think it makes sense to group all the events together?

File: lib/engine/game/g_1880/game.rb [hunk 4]
```diff
@@ -323,25 +329,70 @@ def init_minors
...
+          @log << "-- Event: #{EVENTS_TEXT['all_shares_available'][1]} --"
+          @corporations.each do |c|
+            c.shares.last(5).each { |s| s.buyable = true }
```
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:362]: This is logically correct, but it does mean you have to keep the "magic number" 5 in multiple places. I probably would have done something like `c.shares.each { |s| s.buyable = true}` in order to remove that dependency.

File: lib/engine/game/g_1880/game.rb [hunk 5]
```diff
@@ -323,25 +329,70 @@ def init_minors
...
+          end
+        end
+
+        def event_receive_capital!
+          @log << "-- Event: #{EVENTS_TEXT['receive_capital'][1]} --"
+          @corporations.each do |c|
```
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:368]: Break out into a method that can be called for companies that reach the 5th share after this event occurs.

File: lib/engine/game/g_1880/game.rb [hunk 6]
```diff
@@ -323,25 +329,70 @@ def init_minors
...
+            @bank.spend(amount, c)
+            @fully_funded_corporations << c
+            @log << "#{c.name} recieves #{format_currency(amount)}"
+          end
+          @full_cap_event = true
+        end
+
+        def event_fi_stop_operating!
+          @log << "-- Event: #{EVENTS_TEXT['fi_stop_operating'][1]} --"
+          @minors.each(&:unfloat!)
```
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:381]: The minors still need to check for destination after they stop operating.

File: lib/engine/game/g_1880/game.rb [hunk 7]
```diff
@@ -375,6 +426,29 @@ def train_owner(train)
...
           current_entity
         end
+
+        def round_description(name, round_number = nil)
+          round_number ||= @round.round_num
+          description = "#{name} Round "
+
+          description += @turn.to_s unless @turn.zero?
+          description += ".#{round_number}"
+
```
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:436]: description = super
description += " - Train Marker at..."

File: lib/engine/game/g_1880/game.rb [hunk 8]
```diff
@@ -375,6 +426,29 @@ def train_owner(train)
...
+          description.strip
+        end
+
+        def check_for_foreign_investor_connection(corp)
+          fi = @minors.find { |m| m.shares&.first&.corporation == corp }
+          return true unless fi
+
+          fi_home_token = fi.tokens.first.hex
+
+          graph = Graph.new(self, no_blocking: true)
```
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:448]: Is there a performance difference by creating the graph each time vs storing as an instance var?

File: lib/engine/game/g_1880/game.rb [hunk 9]
```diff
@@ -375,6 +426,29 @@ def train_owner(train)
...
+          description += ".#{round_number}"
+
+          description += " - Train Marker at #{@train_marker.name}" if @train_marker
+
+          description.strip
+        end
+
+        def check_for_foreign_investor_connection(corp)
+          fi = @minors.find { |m| m.shares&.first&.corporation == corp }
+          return true unless fi
```
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:444]: Why true?

File: lib/engine/game/g_1880/game.rb (hunk 3 above)
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb:359]: as in one event_<event> method, but still multiple logging entries?

File: lib/engine/game/g_1880/game.rb [hunk 10]
```diff
@@ -375,6 +426,29 @@ def train_owner(train)
...
+
+          fi_home_token = fi.tokens.first.hex
+
+          graph = Graph.new(self, no_blocking: true)
+          graph.compute(corp)
```
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:449]: graph does lazy evaluation. You don't need to call compute explicitly.

File: lib/engine/game/g_1880/game.rb (hunk 6 above)
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb:381]: The destination check happens on the corp side. After B3, minors/FIs do not operate, no track lay, no train running. The can only be merged to corps when tracks are connected

File: lib/engine/game/g_1880/minor.rb [hunk 11]
```diff
@@ -8,6 +8,18 @@ module Game
     module G1880
       class Minor < Engine::Minor
         include ShareHolder
+        def num_shares_of(corporation, ceil: true)
+          num = percent_of(corporation).to_f / corporation.share_percent
+          ceil ? num.ceil : num
+        end
```
Comment (crericha) [inline lib/engine/game/g_1880/minor.rb:14]: Method is inherited from ShareHolder module. I don't think you need to redefine it here.

File: lib/engine/game/g_1880/game.rb (hunk 10 above)
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb:449]: I copied the code from 48, where i check there to connection between two cities to determine if a corp floats

File: lib/engine/game/g_1880/step/buy_sell_par_shares.rb [hunk 12]
```diff
@@ -113,6 +113,20 @@ def process_par(action)
...
+          def process_buy_shares(action)
+            super
+            extra_cap?(action.entity) if @game.full_cap_event
+          end
+
+          def extra_cap?(corporation)
+            return if corporation.ipo_shares > 5 || @game.fully_funded_corporations.include?(corporation)
+
+            amount = corporation.par_price.price * 5
+            @log << "Fifth share of #{corporation.name} has been bought. #{corporation} recieves #{@game.format_currency(amount)}"
```
Comment (crericha) [inline lib/engine/game/g_1880/step/buy_sell_par_shares.rb:126]: Typo with receives

File: lib/engine/game/g_1880/step/buy_sell_par_shares.rb [hunk 13]
```diff
@@ -113,6 +113,20 @@ def process_par(action)
 
             process_presidents_percent_choice(Action::Choose.new(@parring[:entity], choice: percent_choices.keys.first))
           end
+
+          def process_buy_shares(action)
+            super
+            extra_cap?(action.entity) if @game.full_cap_event
+          end
+
+          def extra_cap?(corporation)
```
Comment (crericha) [inline lib/engine/game/g_1880/step/buy_sell_par_shares.rb:122]: ? methods are not supposed to mutate

Comment (crericha) [reply lib/engine/game/g_1880/step/buy_sell_par_shares.rb:122]: Also put the below functionality in the game class and share it with the capitalization event

Comment (bentziaxl) [reply lib/engine/game/g_1880/step/buy_sell_par_shares.rb:122]: right, good catch, typo. I think it was originally supposed to be a boolean, I switched the method def, but not the name.

File: lib/engine/game/g_1880/game.rb (hunk 9 above)
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb:444]: I don't think i need the check here anyways, the checking step checks that a minor exists before doing the check.

File: lib/engine/game/g_1880/step/buy_train.rb [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+            @round.bought_trains = true
+          end
+
+          def round_state
+            { bought_trains: false }
+          end
+
+          def pass!
+            super
+            @round.bought_trains = false
```
Comment (crericha) [inline lib/engine/game/g_1880/step/buy_train.rb:23]: There is a setup method that gets called after each company is done operating. This is a good place to reset state. https://github.com/tobymao/18xx/blob/master/lib/engine/round/operating.rb#L69

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+module Engine
+  module Game
+    module G1880
+      module Step
+        class CheckFIConnection < Engine::Step::Base
+          ACTIONS = %w[destination_connection].freeze
+
+          def actions(entity)
+            return [] unless entity == current_entity
+            return [] if entity.minor?
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:14]: I'm confused. The check happens on the foreign investors turn. Won't this prevent foreign investors from doing the action?

File: lib/engine/game/g_1880/game.rb (hunk 3 above)
Comment (crericha) [reply lib/engine/game/g_1880/game.rb:359]: multiple event methods, but next to each other in the code.

File: lib/engine/game/g_1880/game.rb (hunk 6 above)
Comment (crericha) [reply lib/engine/game/g_1880/game.rb:381]: > As soon as the two railways have at their disposal a joint network (Station Markers are no obstacle
here) the Foreign Investor is joined to the normal company (this happens at the end of the Foreign
Investor's turn).

File: lib/engine/game/g_1880/step/check_fi_connection.rb (hunk 15 above)
Comment (bentziaxl) [reply lib/engine/game/g_1880/step/check_fi_connection.rb:14]: argh, i misread the rules, I thought it was done at the end of the corp turn. 
however:
"At the end of each turn of each Foreign Investor, it is checked if there is a joint route network (see box) with the associated com"

File: lib/engine/game/g_1880/game.rb (hunk 8 above)
Comment (crericha) [reply lib/engine/game/g_1880/game.rb:448]: I look at the constructor. No logic going on, so what you are doing is fine.

File: lib/engine/game/g_1880/game.rb (hunk 10 above)
Comment (crericha) [reply lib/engine/game/g_1880/game.rb:449]: https://github.com/tobymao/18xx/blob/master/lib/engine/graph.rb#L95

File: lib/engine/game/g_1880/corporation.rb [hunk 16]
```diff
@@ -6,12 +6,16 @@ module Engine
...
     module G1880
       class Corporation < Engine::Corporation
-        attr_accessor :building_permits
+        attr_accessor :building_permits, :full_funded
 
         def floated?
           @floated ||= percent_to_float.zero?
         end
 
+        def fully_funded?
```
Comment (crericha) [inline lib/engine/game/g_1880/corporation.rb:15]: The method doesn't really do anything, so use the attr_accessor instead

File: lib/engine/game/g_1880/game.rb [hunk 17]
```diff
@@ -375,6 +430,23 @@ def train_owner(train)
         def lessee
           current_entity
         end
+
+        def round_description(name, round_number = nil)
+          description = super
+          description += " - Train Marker at #{@train_marker.name}" if @train_marker
+
+          description.strip
```
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:438]: What do you need to strip?

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 18]
```diff
@@ -0,0 +1,128 @@
+# frozen_string_literal: true
+
+require_relative '../../../step/base'
+
+module Engine
+  module Game
+    module G1880
+      module Step
+        class CheckFIConnection < Engine::Step::Base
+          ACTIONS = %w[destination_connection pass].freeze
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:10]: What is the purpose of destination_connection?

Comment (crericha) [reply lib/engine/game/g_1880/step/check_fi_connection.rb:10]: Like, what is the action being used for?

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+
+          def actions(entity)
+            return [] unless entity == current_entity
+            return [] unless entity.minor?
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:14]: If you are going to return no action unless a minor (which makes sense), then the auto action should do the same.

File: lib/engine/game/g_1880/step/check_fi_connection.rb (hunk 18 above)
Comment (crericha) [reply lib/engine/game/g_1880/step/check_fi_connection.rb:10]: Found it below.

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+          def auto_actions(entity)
+            return [] if @merging
+            return [Engine::Action::Pass.new(entity)] unless @game.check_for_foreign_investor_connection(entity)
+
+            [Engine::Action::DestinationConnection.new(entity)]
+          end
+
+          def process_destination_connection(action)
+            corporation = action.entity.shares.first.corporation
+            @merging = [corporation, action.entity]
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:37]: A hash is more maintainable

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+            @game.log << "#{action.entity.full_name} and #{corporation.name} are connected"
+            process_merger
+
+            if action.entity.cash.positive?
+              @state = :choose_precent
+            else
+              process_precent(Action::Choose.new(action.entity,
+                                                 choice: '100%'))
+            end
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:46]: @merging[:state] = :choose_percent
return if action.entity.cash.positive?

process_percent(...)

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+            when :choose_precent
+              process_precent(action)
+            when :choose_token
+              process_token(action)
+            end
+          end
+
+          def setup
+            @state = nil
+            @merging = nil
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:72]: Undefined is nil. No need to set to nil.

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+
+          def process_precent(action)
+            fi = @merging[1]
+            if fi.cash.positive?
+              amount = action.choice == '100%' ? fi.cash : (fi.cash * 20) / 100
+              fi.spend(amount, @merging[0], check_positive: false)
+              @game.log << "#{fi.full_name} transfers #{@game.format_currency(amount)} to #{@merging[0].name}"
+            end
+            process_token(Action::Choose.new(fi, choice: 'Discard')) unless cheapest_unused_token(@merging[0])
+            @state = :choose_token
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:83]: @merging[:state] = :choose_token
return if cheapest_unused_token(...)

process_token(...)

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+            replace_token(fi, fi.tokens.first, cheapest_unused_token(@merging[0])) if action.choice == 'Replace'
+            fi.tokens.first.city.remove_tokens! if action.choice == 'Discard'
+            @game.log << "#{fi.full_name} closes"
+            fi.close!
+
+            pass!
+          end
+
+          def cheapest_unused_token(corp)
+            corp.tokens.find { |t| t.city.nil? }
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:97]: This method isn't deterministic. It's relying on tokens being in cost order, which is true today and probably tomorrow, but not guaranteed.

Comment (crericha) [reply lib/engine/game/g_1880/step/check_fi_connection.rb:97]: corp.tokens.find(&:unused).min_by(&:cost)

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+            city = fi_token.city
+            @game.log << "#{fi.name}'s token in #{city.hex.name} is replaced with a #{corp_token.corporation.name} token"
+            fi_token.remove!
+            city.place_token(corp_token.corporation, corp_token, check_tokenable: false)
+          end
+
+          def choice_name
+            case @state
+            when :choose_precent
+              "Precentage of Foreign Investor's treasury to retain\n"
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:110]: It's more, "Choose destination for Foreign Investor cash"

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+            when :choose_token
+              "Replace Token of Foreign Investor with Corporation?\n"
+            end
+          end
+
+          def choices
+            case @state
+            when :choose_precent
+              ['100%', '20%']
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:119]: ["#{foreign_investor.cash} to #{corporation.name} treasury", "#{foreign_investor.cash * .2} to #{owner.name}"]

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+            fi_token.remove!
+            city.place_token(corp_token.corporation, corp_token, check_tokenable: false)
+          end
+
+          def choice_name
+            case @state
+            when :choose_precent
+              "Precentage of Foreign Investor's treasury to retain\n"
+            when :choose_token
+              "Replace Token of Foreign Investor with Corporation?\n"
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:112]: with corporation token

File: lib/engine/game/g_1880/step/check_fi_connection.rb (hunk 25 above)
Comment (bentziaxl) [reply lib/engine/game/g_1880/step/check_fi_connection.rb:110]: the destination is always the same (the associated corp), the choice is between 100% or only 20% to be moved to the corp

**Comment by crericha** [comment]
Here is the rule on foreign investor cash distribution. It is 20% to owner or 100% to corporation.

> When this happens, the owner of the Foreign Investor decides what to do with the money that has
> been accumulated by the Foreign Investor. It either goes in its entirety to the treasury of the normal
> company with which it is associated, or the owner takes 20% for himself and the remaining 80%
> goes to the Bank.

**Comment by bentziaxl** [comment]
I realized, another rule i read wrongly. The latest change i committed implements this correctly

File: lib/engine/game/g_1880/game.rb [hunk 28]
```diff
@@ -323,16 +360,27 @@ def init_minors
...
         end
 
         def status_array(corporation)
-          return [] unless corporation.floated?
+          return if corporation.minor? || !corporation.floated?
 
           ["Building Permits: #{corporation.building_permits}"]
         end
 
+        def non_floated_corporations
```
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:380]: I don't think this method is used. Do you need it for later?

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 29]
```diff
@@ -0,0 +1,128 @@
...
+          def skip!
+            pass!
+          end
+
+          def process_merger
+            fi = @merging[:fi]
+            share = @merging[:corporation].shares.first
+            @game.share_pool.transfer_shares(share.to_bundle, fi.owner)
+            @game.bank.spend(50, fi.owner)
+            @game.log << "#{fi.owner.name} recieves #{@game.format_currency(50)} and a share of #{share.corporation.name}"
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:56]: typo in receives

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 30]
```diff
@@ -0,0 +1,128 @@
...
+          end
+
+          def process_choose(action)
+            case @merging[:state]
+            when :choose_precent
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:61]: typo in percent. Same is throughout the file.

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 31]
```diff
@@ -0,0 +1,128 @@
...
+          def setup
+            @merging
+          end
+
+          def process_precent(action)
+            fi = @merging[:fi]
+            if fi.cash.positive?
+              amount = action.choice.include?('treasury') ? fi.cash : fi.cash * 0.2
+              destination = action.choice.include?('treasury') ? @merging[:corporation] : current_entity.owner
+              fi.spend(amount, destination, check_positive: false)
```
[the remark this answers is not in the document]
Comment (bentziaxl) [reply lib/engine/game/g_1880/step/check_fi_connection.rb:77]: that was my thought that it doesn't matter, and no need to add a line. Is that ok, or should i change it?

File: lib/engine/game/g_1880/step/check_fi_connection.rb [hunk 32]
```diff
@@ -0,0 +1,128 @@
...
+            @merging[:state] = :choose_token
+            return if cheapest_unused_token(@merging[:corporation])
+
+            process_token(Action::Choose.new(fi, choice: 'Discard'))
+          end
+
+          def process_token(action)
+            fi = @merging[:fi]
+            replace_token(fi, fi.tokens.first, cheapest_unused_token(@merging[:corporation])) if action.choice == 'Replace'
+            fi.tokens.first.city.remove_tokens! if action.choice == 'Discard'
```
Comment (crericha) [inline lib/engine/game/g_1880/step/check_fi_connection.rb:89]: Add a log message

Comment (crericha) [reply lib/engine/game/g_1880/step/check_fi_connection.rb:89]: name chooses to discard the token or something like that

**Review by crericha — APPROVED** [review]

File: lib/engine/game/g_1880/step/check_fi_connection.rb (hunk 31 above)
[the remark this answers is not in the document]
Comment (crericha) [reply lib/engine/game/g_1880/step/check_fi_connection.rb:77]: I think its fine as is

**Pull request merged**